<h1>World Bank International Debt Analysis for 2024</h1> <p>While individuals often take on debt to meet their personal needs, nations may also borrow to bolster their economies. For instance, substantial investments in infrastructure are essential to enhance the quality of life for a country’s citizens. The <a href="https://www.worldbank.org">World Bank</a> serves as a key institution that extends financial assistance to countries in the form of debt.</p> <p>In this notebook, we will explore international debt data compiled by the World Bank, focusing on the debt obligations (in USD) of developing nations across various categories. Our analysis aims to address key questions such as:</p> <ul> <li>What is the overall debt burden borne by the countries included in the dataset?</li> <li>Which country carries the highest debt load, and what is the magnitude of that debt?</li> <li>What is the average debt level across different debt indicators for the countries listed?</li> </ul> <p>

In [ ]:
#Import the required libraries
import pandas as pd
import sqlite3

In [ ]:
#Read the data from our Excel file World Bank Debt
df = pd.read_excel('world_bank_debt.xlsx', sheet_name='Data')
df.head(5)

Country_Name Country_Code Counterpart-Area_Name Counterpart-Area_Code  \
0  Afghanistan          AFG                 World                   WLD   
1  Afghanistan          AFG                 World                   WLD   
2  Afghanistan          AFG                 World                   WLD   
3  Afghanistan          AFG                 World                   WLD   
4  Afghanistan          AFG                 World                   WLD   

                                         Series_Name  Series_Code  1970  1971  \
0  Average grace period on new external debt comm...  DT.GPA.DPPG   NaN   NaN   
1  Average grace period on new external debt comm...  DT.GPA.OFFT   NaN   NaN   
2  Average grace period on new external debt comm...  DT.GPA.PRVT   NaN   NaN   
3  Average grant element on new external debt com...  DT.GRE.DPPG   NaN   NaN   
4  Average grant element on new external debt com...  DT.GRE.OFFT   NaN   NaN   

   1972  1973  ...  2021  2022  2023  2024  2025  2026  2027  2028  2029  2030  
0   NaN   NaN  ...   0.0   0.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1   NaN   NaN  ...   0.0   0.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2   NaN   NaN  ...   0.0   0.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
3   NaN   NaN  ...   0.0   0.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4   NaN   NaN  ...   0.0   0.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[5 rows x 67 columns]

In [ ]:
#Data Preprocessing
print(df.columns.tolist())
year_col = list(range(1970, 2024))
print(year_col)
df = df.drop(columns=year_col)
df = df.drop(columns= list(range(2025, 2031)))
df = df.rename(columns={2024 : 'Debt_Amount'})
df = df.drop(columns = ['Counterpart-Area_Name', 'Counterpart-Area_Code'])
df = df.dropna()

df.head(5)


['Country_Name', 'Country_Code', 'Counterpart-Area_Name', 'Counterpart-Area_Code', 'Series_Name', 'Series_Code', 1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023, 2024, 2025, 2026, 2027, 2028, 2029, 2030]
[1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023]


Country_Name Country_Code  \
98   Afghanistan          AFG   
99   Afghanistan          AFG   
102  Afghanistan          AFG   
103  Afghanistan          AFG   
104  Afghanistan          AFG   

                                           Series_Name     Series_Code  \
98   Debt service on external debt, general governm...  DT.TDS.DEGG.CD   
99   Debt service on external debt, long-term (TDS,...  DT.TDS.DLXF.CD   
102  Debt service on external debt, private nonguar...  DT.TDS.DPNG.CD   
103  Debt service on external debt, public and publ...  DT.TDS.DPPG.CD   
104  Debt service on external debt, public sector (...  DT.TDS.DEPS.CD   

     Debt_Amount  
98   126654245.8  
99   130601245.8  
102    3947000.0  
103  126654245.8  
104  126654245.8

In [31]:
# Create a connection to an SQLite database (this will create a file called 'debt_data.db')
conn = sqlite3.connect('debt_data.db')

# Load the DataFrame into an SQLite table called 'world_debt'
df.to_sql('world_debt', conn, if_exists='replace', index=False)

# Verify the data was loaded by querying the table
query = "SELECT * FROM world_debt LIMIT 5;"
result = pd.read_sql_query(query, conn)
result.head()

Country_Name Country_Code  \
0  Afghanistan          AFG   
1  Afghanistan          AFG   
2  Afghanistan          AFG   
3  Afghanistan          AFG   
4  Afghanistan          AFG   

                                         Series_Name     Series_Code  \
0  Debt service on external debt, general governm...  DT.TDS.DEGG.CD   
1  Debt service on external debt, long-term (TDS,...  DT.TDS.DLXF.CD   
2  Debt service on external debt, private nonguar...  DT.TDS.DPNG.CD   
3  Debt service on external debt, public and publ...  DT.TDS.DPPG.CD   
4  Debt service on external debt, public sector (...  DT.TDS.DEPS.CD   

   Debt_Amount  
0  126654245.8  
1  130601245.8  
2    3947000.0  
3  126654245.8  
4  126654245.8

## 2. Finding the number of distinct countries
<p>From the first 5 rows, we can see the amount of debt owed by <em>Afghanistan</em> in the different debt indicators. But we do not know the number of different countries we have on the table. There are repetitions in the country names because a country is most likely to have debt in more than one debt indicator. </p>
<p>Without a count of unique countries, we will not be able to perform our statistical analyses holistically. In this section, we are going to extract the number of unique countries present in the table. </p>

In [85]:
query = """
SELECT COUNT(DISTINCT Country_name) AS Number_of_Countires 
FROM world_debt;
"""
result = pd.read_sql_query(query, conn)
result.head()

Number_of_Countires
0                  123

## 3. Finding out the distinct debt indicators
<p>We can see there are a total of 123 countries as part of the World Bank Debt Database. As we saw in the first section, there is a column called <code>Series Name</code> that briefly specifies the purpose of taking the debt. Just beside that column, there is another column called <code>Series Code</code> which symbolizes the category of these debts. Knowing about these various debt indicators will help us to understand the areas in which a country can possibly be indebted to. </p>

In [33]:
query = "SELECT DISTINCT Series_Code AS Debt_Indicators FROM world_debt ORDER BY Debt_Indicators;"
result = pd.read_sql_query(query, conn)
result

Debt_Indicators
0         DT.AMT.BLAT.CD
1      DT.AMT.BLAT.GG.CD
2     DT.AMT.BLAT.OPS.CD
3    DT.AMT.BLAT.PRVG.CD
4      DT.AMT.BLAT.PS.CD
..                   ...
141       DT.TDS.PRVT.CD
142    DT.TDS.PRVT.GG.CD
143   DT.TDS.PRVT.OPS.CD
144  DT.TDS.PRVT.PRVG.CD
145    DT.TDS.PRVT.PS.CD

[146 rows x 1 columns]

## 4. Totaling the amount of debt owed by the countries
<p>As mentioned earlier, the financial debt of a particular country represents its economic state. But if we were to project this on an overall global scale, how will we approach it?</p>
<p>Let's switch gears from the series codes now and find out the total amount of debt (in USD in millions) that is owed by all countries in the database. This will give us a sense of how the overall economy of the entire world is holding up.</p>

In [34]:
query = "SELECT ROUND(SUM (debt_amount) / 1000000, 2) AS total_debt_in_million_$ FROM world_debt;"
result = pd.read_sql_query(query, conn)
result

total_debt_in_million_$
0              15043005.87

## 5. Country with the highest debt
<p>Now that we have the exact total of the amounts of debt owed by several countries, let's now find out the country that owns the highest amount of debt along with the amount. <strong>Note</strong> that this debt is the sum of different debts owed by a country across several categories. This will help to understand more about the country in terms of its socio-economic scenarios. We can also find out the category in which the country owns its highest debt. But we will leave that for now. </p>

In [84]:

query = """
SELECT 
    Country_Name, 
    ROUND(SUM(Debt_Amount)/1000000, 2) AS Debt_In_Million$ 
    FROM world_debt 
    GROUP BY Country_Name 
    ORDER BY Debt_In_Million$ DESC 
    LIMIT 1;
"""
result = pd.read_sql_query(query, conn)
result


Country_Name  Debt_In_Million$
0        China        2119167.08

## 6. Average amount of debt across indicators
<p> Having identified China as the country with the highest debt in 2024, a detailed breakdown of China’s debt can be explored <a href="https://datatopics.worldbank.org/debt/ids/country/CHN">here</a>.</p> With a high-level overview of the dataset and its key statistics, we now understand the various debt Series Codes. Let’s dig deeper to find the average debt per country across different indicators in 2024. This analysis will provide insight into the distribution of debt across categories, highlighting which debt types are most significant on average

In [82]:
query = """
SELECT 
    Series_Code, 
    Series_Name, 
    Round(AVG(Debt_Amount)/1000000, 2) as Average_Debt_In_Million$, 
    COUNT(DISTINCT Country_Name) AS Num_Countries 
    FROM world_debt 
    GROUP BY Series_Name, Series_Code 
    ORDER BY Average_Debt_In_Million$ DESC 
    LIMIT 10;
"""
result = pd.read_sql_query(query, conn)
result

Series_Code                                        Series_Name  \
0  DT.TDS.DECT.CD  Debt service on external debt, total (TDS, cur...   
1  DT.TDS.DLXF.CD  Debt service on external debt, long-term (TDS,...   
2  DT.AMT.DLXF.CD  Principal repayments on external debt, long-te...   
3  DT.TDS.DPNG.CD  Debt service on external debt, private nonguar...   
4  DT.TDS.PNGC.CD  PNG, commercial banks and other creditors (TDS...   
5  DT.AMT.DPNG.CD  Principal repayments on external debt, private...   
6  DT.AMT.PNGC.CD  PNG, commercial banks and other creditors (AMT...   
7  DT.TDS.DPPG.CD  Debt service on external debt, public and publ...   
8  DT.TDS.DEPS.CD  Debt service on external debt, public sector (...   
9  DT.TDS.PRVT.CD          PPG, private creditors (TDS, current US$)   

   Average_Debt_In_Million$  Num_Countries  
0                   9104.22            123  
1                   8817.57            123  
2                   6643.72            123  
3                   6385.93             90  
4                   5245.11             88  
5                   5211.88             90  
6                   4342.27             88  
7                   4144.94            123  
8                   4120.84            123  
9                   3245.58             98

## 7. The highest amount of principal repayments
<p>We’ve observed that the indicator <code>DT.AMT.DLXF.CD</code> ranks high in average debt, representing principal repayments on long-term external debt. Countries often incur long-term debt to secure immediate capital, and more details on this category are available <a href = "https://datacatalog.worldbank.org/search/dataset/0038015/International-Debt-Statistics">here</a>.</p> Notably, there’s a significant drop in average debt amounts after the top two indicators, suggesting these are the most critical debt categories. Let’s investigate further to identify the country with the highest principal repayment in the <code>DT.AMT.DLXF.CD</code> category for 2024. This will offer a deeper understanding of that country’s economic challenges, given that economic conditions vary across nations.


In [ ]:
query = """
SELECT 
    Country_Name,
    Series_Code, 
    Series_Name, Round(Debt_Amount/1000000, 2) AS Total_Debt_In_Million$ 
    FROM world_debt 
    WHERE 
        Debt_Amount = 
        (SELECT MAX(Debt_Amount) 
        FROM world_debt 
        WHERE 
        Series_Code = 'DT.AMT.DLXF.CD');
"""
result = pd.read_sql_query(query, conn)
result

Country_Name     Series_Code  \
0        China  DT.AMT.DLXF.CD   

                                         Series_Name  Total_Debt_In_Million$  
0  Principal repayments on external debt, long-te...               136499.27

## 8. The most common debt indicator
<p>China has the highest amount of debt in the long-term debt (<code>DT.AMT.DLXF.CD</code>) category. This is verified by <a href="https://datahelpdesk.worldbank.org/knowledgebase/articles/906522-data-updates-and-errata">The World Bank</a>. It is often a good idea to verify our analyses like this since it validates that our investigations are correct. </p>
<p>We saw that long-term debt is the topmost category when it comes to the average amount of debt. But is it the most common indicator in which the countries owe their debt? Let's find that out. </p>

In [78]:
query = """
SELECT  
    series_code,
    count(Series_code) AS Code_Count
FROM world_debt
GROUP BY Series_Code
ORDER BY Code_Count DESC, series_code
LIMIT 40;
"""
result = pd.read_sql_query(query, conn)
result

Series_Code  Code_Count
0      DT.AMT.BLAT.CD         123
1   DT.AMT.BLAT.PS.CD         123
2      DT.AMT.DEGG.CD         123
3      DT.AMT.DLXF.CD         123
4      DT.AMT.DPPG.CD         123
5      DT.AMT.MLAT.CD         123
6   DT.AMT.MLAT.GG.CD         123
7   DT.AMT.MLAT.PS.CD         123
8      DT.AMT.OFFT.CD         123
9   DT.AMT.OFFT.GG.CD         123
10  DT.AMT.OFFT.PS.CD         123
11     DT.INT.BLAT.CD         123
12  DT.INT.BLAT.PS.CD         123
13     DT.INT.DEGG.CD         123
14     DT.INT.DEPS.CD         123
15     DT.INT.DLXF.CD         123
16     DT.INT.DPPG.CD         123
17     DT.INT.MLAT.CD         123
18  DT.INT.MLAT.GG.CD         123
19  DT.INT.MLAT.PS.CD         123
20     DT.INT.OFFT.CD         123
21  DT.INT.OFFT.GG.CD         123
22  DT.INT.OFFT.PS.CD         123
23     DT.TDS.BLAT.CD         123
24  DT.TDS.BLAT.PS.CD         123
25     DT.TDS.DECT.CD         123
26     DT.TDS.DEGG.CD         123
27     DT.TDS.DEPS.CD         123
28     DT.TDS.DLXF.CD         123
29     DT.TDS.DPPG.CD         123
30     DT.TDS.MLAT.CD         123
31  DT.TDS.MLAT.GG.CD         123
32  DT.TDS.MLAT.PS.CD         123
33  DT.TDS.OFFT.GG.CD         123
34  DT.TDS.OFFT.PS.CD         123
35  DT.AMT.BLAT.GG.CD         122
36  DT.INT.BLAT.GG.CD         122
37  DT.TDS.BLAT.GG.CD         122
38     DT.DIS.DLXF.CD         121
39     DT.DIS.DPPG.CD         121

## 9. Other viable debt issues and conclusion
<p>Our analysis reveals that all 123 countries in the dataset have debt across 35 indicators in 2024, including <code> DT.AMT.DLXF.CD</code>, suggesting a common economic challenge. However, this is only part of the story.</p>
<p>Let’s shift focus from indicators to the total debt amounts and determine the maximum debt each country owes across all categories. This will help identify other plausible economic issues each nation might face. 
<p>In this notebook, we’ve explored global debt patterns for 2024, extracted key statistics, and validated our findings to ensure accuracy, providing a comprehensive overview of the World Bank’s International Debt Statistics

In [75]:
query = """
SELECT 
    Country_Name,
    Country_Code,
    ROUND(SUM(Debt_Amount) / 1000000, 2) AS Total_Debt_In_Million_USD
FROM world_debt
GROUP BY Country_Name, Country_Code
ORDER BY Total_Debt_In_Million_USD DESC
limit 10;
"""
result = pd.read_sql_query(query, conn)
result

Country_Name Country_Code  Total_Debt_In_Million_USD
0               China          CHN                 2119167.08
1          South Asia          SAS                 1539349.49
2              Mexico          MEX                 1061442.01
3           Indonesia          IDN                  976861.13
4               India          IND                  958334.25
5              Brazil          BRA                  931575.26
6             Turkiye          TUR                  711042.54
7  Russian Federation          RUS                  665263.72
8    Egypt, Arab Rep.          EGY                  434825.37
9            Colombia          COL                  280572.21